src/nidcpower/system_tests/test_system_nidcpower.py

```diff
@@ -1063,6 +1065,50 @@ def test_get_lcr_compensation_last_date_and_time(self, session, compensation_typ
         assert last_compensation_datetime.hour == 0
         assert last_compensation_datetime.minute == 0
 
+    # Multi-Threading tests
+    def test_multi_threading_lock_unlock(self, session):
+        release_lock = threading.Event()
+
+        def lock_wait_unlock():
+            session.lock()
+            release_lock.wait()
+            session.unlock()
+
+        def lock_unlock():
+            session.lock()
+            session.unlock()
+
+        # test that lock, unlock functions work properly
+        # No need to test locking for driver, but the gRPC_interpeter doesn't use the driver to lock
+        t1 = threading.Thread(target=lock_wait_unlock)
+        t2 = threading.Thread(target=lock_unlock)
+
+        t1.start()
+        t2.start()
+
+        t1.join(0.5)
+        time.sleep(0.5)
+        # t1 is blocked by the event, t2 should be blocked by t1's lock
+        assert t2.is_alive()
+        release_lock.set()
+        t2.join(0.5)
```

Comment:
don't use a timeout here, or if you insist, make it enormous like 10 s

Reply:
It feels a little uncomfortable setting no timeout for something that might block forever, but I suppose since in my testing a timeout didn't keep _LockSession from blocking, there isn't much point.

Reply:
Removed join timeouts, increased sleep time and added another threading.Event.